minigui/app.ts

```diff
@@ -0,0 +1,194 @@
+  search: string[] | Move[];
+  n: number[];
+  dq: number[];
+  pv?: string[] | Move[];
+}
+
+interface GameStateMsg {
+  board: string;
+  toPlay: string;
+  lastMove?: string;
+  moveNum: number;
+  q: number;
+  gameOver: boolean;
+}
+
+// App base class used by the different Minigui UI implementations.
+abstract class App {
+  // WebSocket connection to the Miniui backend server.
+  protected gtp = new Socket();
+
+  // Board size.
+  protected size: number;
+
+  // True if Minigo is processing a genmove command.
+  protected engineBusy = false;
+
+  // The list of board positions played.
+  // TODO(tommadams): This will need to be turned into a tree when we add an
+  // analysis mode.
+  protected positionHistory: Position[];
+
+  // The current position being displayed on the boards.
+  protected activePosition: Position;
+
+  // True when the backend reports that the game is over.
+  protected gameOver = true;
+
+  // Whose turn is it.
+  protected toPlay = Color.Black;
+
+  // List of Board views.
+  private boards: Board[] = [];
+
+  protected abstract onGameOver(): void;
+
+  constructor() {
+    this.gtp.onData('mg-search', this.onSearch.bind(this));
+    this.gtp.onData('mg-gamestate', this.onGameState.bind(this));
+  }
+
+  protected connect() {
+    let uri = `http://${document.domain}:${location.port}/minigui`;
+    return this.gtp.connect(uri).then((size: number) => {
+      this.size = size;
+    });
+  }
+
+  protected init(boards: Board[]) {
+    this.boards = boards;
+  }
+
+  protected newGame() {
+    this.activePosition = new Position(
+      0, util.emptyBoard(this.size), null, Color.Black)
+    this.positionHistory = [this.activePosition];
+
+    this.gtp.send('clear_board');
+    this.gtp.send('gamestate');
+    this.gtp.send('info');
+
+    // Iterate over the data-* attributes attached to the main minigui container
+    // element, looking for data-gtp-* attributes. Send any matching ones as GTP
+    // commands to the backend.
+    let containerElem = document.querySelector('.minigui') as HTMLElement;
+    if (containerElem != null) {
+      let dataset = containerElem.dataset;
+      for (let key in dataset) {
+        if (key.startsWith('gtp')) {
+          let cmd = key.substr(3).toLowerCase();
+          let args = dataset[key];
+          this.gtp.send(`${cmd} ${args}`);
+        }
+      }
+    }
+
+    this.updateBoards(this.activePosition);
+  }
+
+  protected updateBoards(state: any) {
+    for (let board of this.boards) {
+      if (board.update(state)) {
+        board.draw();
+      }
+    }
+  }
+
+  protected onSearch(msg: SearchMsg) {
+    // Parse move variations.
+    msg.search = util.parseMoves(msg.search as string[], this.size);
+    msg.toPlay = util.parseGtpColor(msg.toPlay as string);
+    if (msg.pv) {
+      msg.pv = util.parseMoves(msg.pv as string[], this.size);
+    }
+
+    // Update the board state with contents of the search.
+    const props = ['n', 'dq', 'pv', 'search'];
```

Comment:
Adding more would require updating this list, yes. We'd pretty much have to do that anyway if we want to send extra properties from the backend because the current structures are statically typed.

I guess we could make the SearchMsg and PositionHistory hold arbitrary key/value pairs and rely on the UI to display them correctly. That's an interesting idea, so I've added a TODO for now.